### ISSUE TITLE
Confine on dispatch queue rather than thread

### ISSUE THREAD

**Issue by rogernolan** [realm/realm-cocoa#1097]
tl;dr Realm uses thread ids and throws exceptions when they change.  Checking thread ID is not acceptable under GCD.

Realm docs [state](http://realm.io/docs/cocoa/0.87.1/#background-operations) objects are not gcd queue safe and when running under GCD:

> you must get an RLMRealm instance in each thread/dispatch_queue in which you want to read or write

I do this:

```
    // Local copy of the persisted task valid for the worker queue.
__block OFEPersistedTask* taskCopyOnWorkerQueue = nil;

NSString* taskID = aTask.taskID;
dispatch_sync(_workerQueue, ^{
    NSPredicate *pred = [NSPredicate predicateWithFormat:@"taskID = %@", taskID];
    RLMResults *results = [OFEPersistedTask objectsWithPredicate:pred];
    taskCopyOnWorkerQueue = [results firstObject];
}
```

To get a copy of my Task (RLMObject subclass) on the correct queue. Unfortunately, in `RLMVerifyAttached` Realm checks the thread ID through `RLMCheckThread` (i.e. not queue) which does not work under GCD.

A GCD queue can run its tasks on many different threads - thats kindof the point. From the [Apple docs](https://developer.apple.com/library/ios/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html):

> The currently executing task runs on a distinct thread **(which can vary from task to task)**

My **emphasis**... So, even if I grab a RLMObject on my queue I still risk getting killed (actually, no risk, I do get killed) later for running on the wrong thread.

The only workaround I can think of is to store a primary key with __block scope and then perform another Realm fetch at the beginning of every task. That feels so suboptimal I don't think you can claim Realm is compatible with GCD _at all_.

**Comment by vittoriom** [comment]
I also encountered this issue where I get a RLMObject in one thread (not relevant) and then I start some work on another dispatch_queue and I get "Realm accessed from incorrect thread" when reading values of the object. It would be so nice to have a "queue-safe read-only copy or view of the object" here... 
Any solution?

**Comment by ebluehands** [comment]
I also encountered this issue and would be very interested in a solution.

**Comment by jpsim** [comment]
@vittoriom and @ebluehands: we're actively working on this functionality in the form of "thread handoff": e.g. run a query which returns objects on separate threads. We'll post to this issue once we have more to share.

In the mean time, you can work around most of these issues by passing around primary keys to objects between threads rather than the objects themselves (similar to using `NSManagedObjectID`s in Core Data). This does mean that you'll have to re-fetch the object on the target thread, which we understand is sub-optimal, which is why we're actively working on thread handoff.

**Comment by vittoriom** [comment]
@jpsim thanks for the follow up, I did fetch the object on the target thread indeed, and for now it will do the job. But it's in the middle of a complex async set of blocks dispatched in a dispatch group so it would look much cleaner if I could just share the object (in read-only mode) on other queues as well.
I'll watch for more comments on this issue! Have a good weekend

**Comment by rciobanu-omnifone** [comment]
@jpsim, @tgoyne, what about entities that don't have a primary key defined? AFAIK, `NSManagedObjectID` exists for every `NSManagedObject` once it's been added to the DB. Primary Keys, on the other hand, are optional. In some cases it's even recommended against having PKs for Realm entities: https://github.com/realm/realm-cocoa/issues/1234#issuecomment-67192865

**Comment by jpsim** [comment]
> In some cases it's even recommended against having PKs for Realm entities: #1234 (comment)

That comment was from a while ago. Primary keys are supported and you're encouraged to use them in cases in which you need a unique identifier for your objects.

We've opted against going the Core Data route of unconditionally adding an additional column on each model for the sake of global object ID's, but you're free to do so yourself.

The thread handoff solution we're working on will support all objects, not just ones with primary keys.

**Comment by tettoffensive** [comment]
I'm a bit confused about what the current status of this is? I've been using realm with GCD by grabbing the id of the object before the dispatch and then querying for it on that thread. But this can get messy and sometimes I don't have the id. Plus, this can get quite messy keeping track of which thread I'm on. What's the current best practice for using realm on another thread to keep it async from the UI?

**Comment by vittoriom** [comment]
Not sure it's a best practice, but what I'm doing in the project I'm working on is:

- Persistence layer has a reference to a "realm factory". In the simplest case, it just returns <code>RLMRealm.defaultRealm</code> everytime. (Added bonus, in integration tests I can inject a "test factory" that reads and writes to a bundled database with known content)
- Persistence layer driver begins every read/write on the persistence layer with a <code>writeInTransaction</code> or <code>readInTransaction</code> call. This internally just dispatches the closure argument to a background queue, or calls <code>transactionWithBlock</code> on the realm.
- All the reads and writes done in the persistence layer reference the factory-generated realm (so there's no problem of accessing a realm on the wrong thread at any time)
- Immediately after reading some data, the persistence layer converts the <code>RLMObject</code> object to a lightweight struct so that it can be exposed to outer layers and passed around threads without any danger

Hope it helps :)

**Comment by yoshyosh** [comment]
@tettoffensive I think the easiest way is to call your realm on each thread. `let realm = RLMRealm.defaultRealm()`, then fetch the objects you want on that thread. If you want to pass objects, you can create an NSArray of primary keys and then fetch them again from the main thread. For the most part, Realm is really fast to be run on the main thread, so you don't have to worry about performance issues

**Comment by rciobanu-omnifone** [comment]
@yoshyosh the problem with this approach is that one would usually need the objects corresponding to the IDs in the same order that the IDs were, which is not currently available. There is no ```+ (instancetype)objectsInRealm:(RLMRealm *)realm forPrimaryKeys:(NSArray *)primaryKeys```
Example: if the heavy lifting is done in a background thread and as part of that work the results are being ordered. Then, on the main thread, I get an array of IDs which I want to expand to full objects. And I believe, more often than not, I'd want the expanded objects to be in the exact same order.
See https://github.com/realm/realm-cocoa/issues/1325
I.E. you would have to re-order the results on the UI thread or fetch them one by one and assemble an array of results.

**Comment by jasperblues** [comment]
I wanted to use the DAO pattern with Realm to encapsulate common persistence operations. 

* Create a protocol that defines createOrUpdate, findBy, listBy, delete .  .
* Create a Realm implementation

It didn't go well because the returned results would be accessed from another thread. I'd need either: 

* Copy to another thread ability. 
* Ability to 'detach' objects from Realm to behave like regular in-memory objects, and subsequently reattach (have seen this in some ORMs in other languages - maybe adds unwanted complexity, but just mentioning).

**Comment by kunalsood** [comment]
It's been a while since there has been any discussion on this, so, here's me adding my +1

**Comment by segiddins** [comment]
We just merged support for delivering results from one thread to another in the underlying database engine, and I've been working on exposing that in the Cocoa binding.

**Comment by zdavison** [comment]
@vittoriom

> Immediately after reading some data, the persistence layer converts the RLMObject object to a lightweight struct so that it can be exposed to outer layers and passed around threads without any danger

Curious as to what this looks like in practice? IMO your solution is the ideal one, but do you store non-ORM blobs in Realm, or do you have two forms of each Model, one that subclasses Realm and one that doesn't (the latter being your immutable struct)

**Comment by vittoriom** [comment]
@zdavison 

The latter option, we have two "forms" (I like to call them "views", since for some complex cases we have even more than 2) of each model. We have a subclass of `RLMObject` for our persistence layer and a simple immutable `struct` to pass around threads and functions. Also, we have an object that takes care, given an object of type 1 (subclass of `RLMObject`), to return an immutable `struct` of type 2. This way we leave the two things very dummy and passive (plus, immutable structs have a number of advantages), and we keep the knowledge to convert one to another in a separate testable object.

I know it may look like a lot of work for every new model object, but eventually it's mechanic enough that one can do it in a few minutes and it brings simplicity and advantages that cannot be beaten IMHO. Also, as I wrote before, for very complex model objects (with that I mean objects that have a high number of relations and nested and nested children), we have not one but two struct "views", a lightweight one to pass around when you need just some kind of "metadata" and the full one when you need to access every bit of information :)

**Comment by jasperblues** [comment]
Not sure if this idea makes sense as I haven't been using Realm for long, but would it make sense to implement `detach` and `attach` methods? Where detach makes the object into a PONSO (plain old Objective-C object) and attach reattaches to Realm. So if I'm about to call a block: 

```objc
[myRealmObject detach];
[myService doSomethingWithResult:^(id result) { 
   [myRealmObject attach]; //Reattaches using primary key
       //continue
   }
```

When detaching there'd need to be rules around eagerly loading related objects/collections: 

- Throw an error
- Perhaps allow detach and loading specified related entities eagerly. 

Again apologies, if I've misunderstood something as I'm new to Realm.

**Comment by jpsim** [comment]
You can essentially do that now by passing a persisted object to an RLMObject initializer, initializing an in-memory object with the value of a persisted one, or vice-versa.

**Comment by jasperblues** [comment]
@jpsim Awesome - can you point me at a code sample?

**Comment by jpsim** [comment]
`Person *inMemoryPerson = [[Person alloc] initWithValue:persistedPerson];`

**Comment by jasperblues** [comment]
Thanks @jpsim . . that looks ok. What I was doing before was: 
```
NSString *uuid = person.primaryKey
[service doSomethingOnNewThread:^{ 
    person* person = [Person objectWithPrimaryKey:uuid];
    //continue
   }
```

**Comment by jpsim** [comment]
If all you need to do is re-fetch an object, your approach is preferable. The reason Realm doesn't do this automatically for you is that the object may have changed between those two threads, so keep that in mind.

**Comment by jasperblues** [comment]
Thank you @jpsim

**Comment by siuying** [comment]
@jpsim do you mean the in-memory object is safe across thread?

**Comment by jpsim** [comment]
> @jpsim do you mean the in-memory object is safe across thread?

Standalone RLMObjects are just plain old NSObjects, with the same threading implications as if you were inheriting from NSObject.

**Comment by anlaital** [comment]
@jpsim should creating a standalone object by using initWithValue work also if the object has a list of persisted child objects? Those don't seem to be created standalone, or is there some trick to it?

**Comment by segiddins** [comment]
@anlaital no, that would've been implemented via https://github.com/realm/realm-cocoa/pull/2043 but we decided not to merge it

**Comment by anlaital** [comment]
@segiddins what's the recommended approach for detaching a complex persisted object and doing  changes on it that may or may not be persisted in the future?

**Comment by segiddins** [comment]
@anlaital you'll have to do your own recursive traversal of the object graph, creating standalone copies of every object you encounter -- if you really do want a 'detached' copy of your object graphs

**Comment by jrpinteno** [comment]
Hi, I just wanted to ask how is this going on. As it was stated in some other open issue now closed, relying on fetching with primary keys is far from from optimal. Specially since every time that fetch is performed, a new instance is created.

Do you have any update on the work you're doing for native cross-thread object sharing/passing?

**Comment by jpsim** [comment]
We're working on this. There's some major architectural work needed for this to happen.

**Comment by hyouuu** [comment]
Thanks for the update - just want to upvote priority for this issue as well. I recently decided to migrate to Realm, and this made me have to do all the work in the main queue because that's the only queue guaranteed to be executed on the same main thread. Have to do lots of Object(value:) copies for background stuff, and had to change from reference to just store IDs - the lack of this ability basically changed how I lay out the whole data flow.

**Comment by siuying** [comment]
I'm currently making immutable copy of view objects (as Swift struct) that is safe to pass across threads. Built in ways to handle cross thread use case would definitely helpful, as we have more cores than ever now on iOS.

**Comment by ReDetection** [comment]
@hyouuu just wanted to ask, what is the problem of executing realm queries on some particular background thread and do these `Object(value:)` copies to deliver objects to the main thread instead?

**Comment by Kirow** [comment]
@ReDetection if you will execute queries using GCD - this will not guarantee single thread. If you will use nested NSThread for this task - this will be much more slower. So I think GCD serial queue support is highly required. In fact if I remove exception from realm code, it works fine, but I were told not to do that.

**Comment by hyouuu** [comment]
@ReDetection - @Kirow explained it well - dispatch to a particular queue doesn't guarantee it will be the same thread, except for the main queue

**Comment by siuying** [comment]
@hyouuu after I migrated to Realm, I refactor lots of code from passing objects across thread to passing object ID instead. It's a bit annoying as many of those code now coupled with Realm fetch and save logic. But you can safely offload heavy tasks to other queues.

**Comment by hyouuu** [comment]
@siuying initially I kept my parent - child using List<Object> relationships, thus the problem seems unsolvable with background thread - say on Thread1 my Parent1 & its Child1 were added to Realm, then on Thread2 I use Parent1's ID to fetch Parent1, then it will fail because Child1 was not stored on Thread2. Not sure if I used it incorrectly though.

I did change that so my Parent1 only stores children IDs in a comma separated string (no string array support should be another hi-pri issue to work on...), so background thread might be ok to work again, but I'd rather wait for this feature to be completed than hacking up a complicated logic for temporary usage...

**Comment by borut-t** [comment]
I am also waiting for solution ...

**Comment by nguyenchiencong** [comment]
:+1:

**Comment by dltlr** [comment]
+1

**Comment by narciero** [comment]
+1

**Comment by siuying** [comment]
After writing so many code to work around thread issue, I wish Realm can just use immutable data type, where I can just send data freely.

**Comment by Kirow** [comment]
Any updates?

**Comment by jpsim** [comment]
> Any updates?

We're tracking progress in this issue. There's nothing new to report other than we've been working on a few different designs that could ensure both safety *and* actually be useful, but we're still in the design phase.

We're also tracking a more specialized form of object thread handoff in #3136.

**Comment by Kirow** [comment]
I have application with complex filtering functionality. One filter change will run few queries.
Queries do not take a lot of time, 0.2 - 5 sec to execute  them all (depends of complexity). But it is too much to execute them in main thread, so I run these queries in background and pass results to main thread. The most obvious way to do that - enumerate all RLMObjects in RLMResults and create array of primary keys that can be used in another thread.

But I got awful results on iPod5. It takes more than 10seconds to enumerate all of my objects an create array of objects (100k+ objects). And such solution will not allow me to apply filter on previous result, I will need to execute it from scratch.

Do you have any advice for me?

**Comment by jpsim** [comment]
If you don't hold on to an `RLMResults` by passing it directly to `-[RLMResults addNotificationBlock:]`, the query will never be executed on the main thread, it will only ever be executed in the background, and the notification block will then hand over those results back to the main thread:

```objc
// on main thread
self.results = nil;
self.token = [[MyModel objectsWhere:someExpensivePredicateQuery] addNotificationBlock:^(RLMResults<MyModel *> *results, NSError *error) {
    // back on the main thread
    self.results = results;
    [self.myViewController updateUI];
}];
```

**Comment by Kirow** [comment]
@jpsim , thanks a lot. This seems to be working very well with first test sample. I will report about results after rewrite all my code.

**Comment by RashmiZK** [comment]
Are there any updates on this issue?

**Comment by jpsim** [comment]
This thread ended up encompassing several different topics, questions and scenarios. @RashmiZK can you be more specific about what you'd like clarification or updates on?

**Comment by RashmiZK** [comment]
Thank you for quick response, appreciate it.

With reference to link - https://github.com/realm/realm-cocoa/issues/2846, I 
would like to know updates on – fetching Realm objects on background thread 
and passing it to main thread.



Regards,

Rashmi

**Comment by jpsim** [comment]
That'd require a thread handover mechanism, which is in progress at #3946.

**Comment by austinzheng** [comment]
We have a handover mechanism now. This ticket is too far ranging to be of any further use.
